### DISCUSSION TITLE
The ability to pass down any method as method argument

### DISCUSSION THREAD

**Discussion by Job-Bouwhuis** [dotnet/csharplang#6890] [Language Ideas]
whether the method has 0 or 16 arguments shouldn't matter in this feature.

Why is this useful? 
The ability to pass any method as an argument to another method allows for less code, and cleaner code.
A suggestion on how to make this would be to be able to get MethodInfo directly on a method. be it implicitly or explicitly.

**Comment by Richiban** [comment]
Okay, but what would you do with a reference to a method that takes an unknown number of arguments and returns an unknown result? You couldn't call it, so why do you want to be able to pass it?

**Reply by Job-Bouwhuis** [reply]
right now, if i want to pass in any method, i need to make overloads with delegates, Action, or Func that accepts different amount of arguments. you can find the number of parameters in the MethodInfo, so there you could take that information and fetch the arguments to use some other way. also, it could then be used for things like seeing in what type the method is implemented, or other behavior that is possible with the MethodInfo class, without having to have a million overloads with delegates, Action, or Func accepting the different amounts of arguments

**Reply by Richiban** [reply]
If you're already using MethodInfo (which is part of reflection) then why not just call the method through reflection as well? I'm not really sure what language feature would make this easier for you, other than perhaps shortening the syntax a little.

Here's an example:

```cs
using System;

public class C 
{
    public void M() 
    {
        Console.WriteLine("Inside M");
    }    
}

public static class Program
{
    public static void Main()
    {
        var methodInfo = typeof(C).GetMethod("M");
        
        methodInfo.Invoke(new C(), new object[] {});
    }
}
```

What more do you need?

**Reply by Job-Bouwhuis** [reply]
i mean, sure this is possible, however it searches for the method using a string... it would be way nicer to have a way to get the MethodInfo directly from the method instead of having to search for it every time, or have it be converted another way instead of always having to do the search on the type. do you get what im saying?

**Reply by Richiban** [reply]
You could introduce a `MethodWrapper` class that wraps all the various overloads of `Func` / `Action`. That way you've still got to handle lots of variants of `Func` etc but you only have to do it in one place.

I'm saying that this is a pretty niche scenario you have here, and since it _is_ currently possible it's unlikely to get its own language feature, especially since what you're asking for is inherently not type-safe (you're resolving a method at runtime).

**Reply by Job-Bouwhuis** [reply]
hmmm okay then... quite unfortunate. ill create a method wrapper class and attempt to create this myself. thanks for making time for my suggestion! have a nice day

**Reply by HaloFour** [reply]
This is supported if the method accepts `Delegate`.  The compiler will use the existing `Action<...>`/`Func<...>` delegates for methods that do fit those signatures, otherwise the compiler will emit it's own delegate type.

[SharpLab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJ2WMCAbJmAHbCYBiA9swBR0MCGCANDfUxc4/TkPSjBXACyTuAVjlDKSrgHZVADl7FSe/QcNiuATlUIADOb4DuCEbaEIJjnrNcJFHlbsN//xggaHtoeZq5wFgCUmAC8AHyYFgDcvphpGNhU2NKYACoApgDOwGwxRCh+ALJsuAB0LMxRqZWkAL5pGeTZMLk1ACIFEAUA5lzABZgAJuWoaR0obUA=)

**Comment by Xyncgas** [comment]
what about ```dynamic```
don't worry, JIT optimizes once it understands what method was passed as argument once dynamic is resolved, and replacing the old code with new optimized branch via hot swapping
